File: hook.go
package main

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"sync"
	"sync/atomic"
	"syscall"
	"time"

	"codeberg.org/miekg/dns"
	"github.com/kkkgo/mini-ppdns/mlog"
	"github.com/kkkgo/mini-ppdns/pkg/cache"
)

const (
	hookCheckTimeout = 30 * time.Second
	hookCmdTimeout   = 60 * time.Second
	// maxHookOutput caps how much stdout+stderr we retain from a hook's
	// check command. A runaway hook producing unbounded output cannot grow
	// the buffer past this size; excess bytes are dropped on the floor so
	// matching only sees the first maxHookOutput bytes.
	maxHookOutput = 64 * 1024
)

// boundedBuffer is a bytes.Buffer wrapper that stops accepting data once
// its contents reach max. Writes still report full consumption so exec's
// io-copy loop exits cleanly. Safe for single-writer use; exec.Cmd
// serializes writes when Stdout == Stderr, which is the only usage here.
type boundedBuffer struct {
	buf bytes.Buffer
	max int
}

func (w *boundedBuffer) Write(p []byte) (int, error) {
	if w.buf.Len() >= w.max {
		return len(p), nil
	}
	if room := w.max - w.buf.Len(); len(p) > room {
		w.buf.Write(p[:room])
		return len(p), nil
	}
	return w.buf.Write(p)
}

func (w *boundedBuffer) Bytes() []byte { return w.buf.Bytes() }

type hookMonitor struct {
	cfg      *HookConfig
	failed   *atomic.Bool
	logger   *mlog.Logger
	dnsCache *cache.Cache[CacheKey, *dns.Msg]

	// shutdownCtx is cancelled when the program begins to shut down. Hook
	// command goroutines watch it so they cannot outlive the main loop.
	shutdownCtx context.Context

	// cmdWg tracks in-flight switch-exec goroutines so main can wait for
	// them to finish before closing the logger.
	cmdWg sync.WaitGroup
	// cmdMu guards cmdClosed. Pairing every cmdWg.Add with the cmdClosed
	// check under this lock prevents the run loop from calling Add(1)
	// concurrently with the cmdWg.Wait() in Wait() — that race is a
	// sync.WaitGroup misuse panic that could fire during shutdown when a
	// check happens to succeed right after shutdownCtx is cancelled.
	cmdMu     sync.Mutex
	cmdClosed bool
}

// cmdAdd registers one tracked command goroutine. It returns false once
// Wait() has been called (shutdown started), in which case the caller must
// not start the goroutine. Holding cmdMu across the cmdClosed check and the
// Add guarantees Add never races the Wait inside Wait().
func (hm *hookMonitor) cmdAdd() bool {
	hm.cmdMu.Lock()
	defer hm.cmdMu.Unlock()
	if hm.cmdClosed {
		return false
	}
	hm.cmdWg.Add(1)
	return true
}

// Wait blocks until all hook-spawned command goroutines have returned.
// After Wait returns (or even begins), no further tracked goroutines can
// start — cmdAdd will refuse them.
func (hm *hookMonitor) Wait() {
	hm.cmdMu.Lock()
	hm.cmdClosed = true
	hm.cmdMu.Unlock()
	hm.cmdWg.Wait()
}

// runCmdTracked runs cmdStr while holding a reference in cmdWg so that
// graceful shutdown can wait for it. No-op once shutdown has started.
func (hm *hookMonitor) runCmdTracked(cmdStr string) {
	if cmdStr == "" {
		return
	}
	if !hm.cmdAdd() {
		return
	}
	go func() {
		defer hm.cmdWg.Done()
		runExecCmd(hm.shutdownCtx, hm.logger, cmdStr)
	}()
}

func (hm *hookMonitor) run(ctx context.Context) {
	sleepTime := time.Duration(hm.cfg.SleepTime) * time.Second
	retryTime := time.Duration(hm.cfg.RetryTime) * time.Second
	failCount := 0
	wasDown := false

	for {
		err := hm.check(ctx)
		if err == nil {
			failCount = 0
			if wasDown {
				hm.failed.Store(false)
				wasDown = false
				hm.logger.Infow("[hook] main DNS recovered, switching back to main DNS")
				hm.runCmdTracked(hm.cfg.SwitchMainExec)
			}
			select {
			case <-ctx.Done():
				return
			case <-time.After(sleepTime):
			}
		} else {
			failCount++
			hm.logger.DebugEventw("[hook] check failed",
				mlog.Int("count", failCount),
				mlog.Int("threshold", hm.cfg.Count),
				mlog.Err(err))
			if failCount >= hm.cfg.Count && !wasDown {
				hm.failed.Store(true)
				wasDown = true
				// Clear DNS cache so stale main-DNS results are purged immediately
				if hm.dnsCache != nil {
					hm.dnsCache.Flush()
					hm.logger.Infow("[hook] DNS cache flushed")
				}
				hm.logger.Warnw("[hook] main DNS marked DOWN, switching to fallback",
					mlog.Int("failures", failCount))
				// Wait retryTime before executing switch_fall_exec so fallback DNS
				// is active and available for the script (e.g. sending notifications).
				// cmdAdd (not a bare cmdWg.Add) so this can't race main's Wait().
				if hm.cfg.SwitchFallExec != "" && hm.cmdAdd() {
					go func(cmd string) {
						defer hm.cmdWg.Done()
						select {
						case <-ctx.Done():
							return
						case <-time.After(retryTime / 2):
							if !hm.failed.Load() {
								return
							}
							// Run inline: we are already in a dedicated
							// goroutine for the delay, so runOptionalCmd's own
							// `go` wrapper would just add another hop.
							runExecCmd(hm.shutdownCtx, hm.logger, cmd)
						}
					}(hm.cfg.SwitchFallExec)
				}
			}
			select {
			case <-ctx.Done():
				return
			case <-time.After(retryTime):
			}
		}
	}
}

// shellCommand creates an exec.Cmd that runs cmdStr through the system shell.
// Uses SHELL env var (fallback /bin/sh). The shell is started in a new
// process group (Setpgid) so that when ctx is cancelled, the custom
// Cancel handler can SIGKILL the entire group with kill(-pgid). Without
// this, exec.CommandContext only kills the direct shell child, leaving
// any backgrounded grandchildren ("&" jobs in scripts) as orphans that
// outlive the resolver and stall shutdown.
func shellCommand(ctx context.Context, cmdStr string) *exec.Cmd {
	shell := os.Getenv("SHELL")
	if shell == "" {
		shell = "/bin/sh"
	}
	cmd := exec.CommandContext(ctx, shell, "-c", cmdStr)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	cmd.Cancel = func() error {
		// cmd.Process is non-nil whenever Cancel runs (CommandContext
		// only fires Cancel after Start). Send SIGKILL to -pid which
		// the kernel interprets as the process group leader's group.
		if cmd.Process != nil {
			return syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
		}
		return nil
	}
	// WaitDelay caps how long Wait blocks after Cancel before forcibly
	// closing copy goroutines. Five seconds is comfortably above any
	// realistic SIGKILL propagation time but short enough that a stuck
	// grandchild's open stdout pipe doesn't extend shutdown indefinitely.
	cmd.WaitDelay = 5 * time.Second
	return cmd
}

func (hm *hookMonitor) check(ctx context.Context) error {
	checkCtx, cancel := context.WithTimeout(ctx, hookCheckTimeout)
	defer cancel()

	cmd := shellCommand(checkCtx, hm.cfg.Exec)
	sink := &boundedBuffer{max: maxHookOutput}
	cmd.Stdout = sink
	cmd.Stderr = sink
	err := cmd.Run()
	output := sink.Bytes()

	exitCode := 0
	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			exitCode = exitErr.ExitCode()
		} else {
			// Command failed to run (not found, permission denied, timeout, etc.)
			return fmt.Errorf("execution failed: %v", err)
		}
	}

	// If neither exit_code nor keyword is configured, default: success = exit code 0
	if !hm.cfg.ExitCodeSet && hm.cfg.Keyword == "" {
		if exitCode == 0 {
			return nil
		}
		return fmt.Errorf("exit code %d (expected 0)", exitCode)
	}

	if hm.cfg.ExitCodeSet && exitCode != hm.cfg.ExitCode {
		return fmt.Errorf("exit code %d (expected %d)", exitCode, hm.cfg.ExitCode)
	}
	if hm.cfg.Keyword != "" && !bytes.Contains(output, []byte(hm.cfg.Keyword)) {
		return fmt.Errorf("output does not contain keyword %q", hm.cfg.Keyword)
	}
	return nil
}

// runExecCmd synchronously executes cmdStr with the hook command timeout,
// logging any failure. Callers already running in a goroutine should prefer
// this to avoid an unnecessary extra goroutine hop. If parentCtx is cancelled
// (shutdown), the spawned process is killed so it cannot outlive main.
func runExecCmd(parentCtx context.Context, logger *mlog.Logger, cmdStr string) {
	if cmdStr == "" {
		return
	}
	if parentCtx == nil {
		parentCtx = context.Background()
	}
	ctx, cancel := context.WithTimeout(parentCtx, hookCmdTimeout)
	defer cancel()
	cmd := shellCommand(ctx, cmdStr)
	out, err := cmd.CombinedOutput()
	if err != nil {
		logger.Warnw("[hook] switch exec failed",
			mlog.String("cmd", cmdStr),
			mlog.Err(err),
			mlog.String("output", string(out)))
	}
}

File: src/hook.rs
// Copyright (c) 2026, https://blog.03k.org. All rights reserved.

//! Hook health monitor.
//!
//! Periodically runs a shell probe of the primary DNS's health. After `count`
//! consecutive failures it marks the primary down (flipping the shared flag the
//! handler reads to force fallback), flushes the cache, and runs the optional
//! `switch_fall_exec`. On recovery it flips back and runs `switch_main_exec`.

use std::process::Stdio;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use tokio::io::AsyncReadExt;
use tokio::process::Command;
use tokio::sync::watch;

use crate::cache::Cache;
use crate::config::HookConfig;
use crate::pplog::{Reporter, SEVERITY_INFO, SEVERITY_WARN};

const HOOK_CHECK_TIMEOUT: Duration = Duration::from_secs(30);
const HOOK_CMD_TIMEOUT: Duration = Duration::from_secs(60);
const MAX_HOOK_OUTPUT: usize = 64 * 1024;

pub struct HookMonitor {
    pub cfg: HookConfig,
    pub failed: Arc<AtomicBool>,
    pub cache: Arc<Cache>,
    pub pplog: Option<Arc<Reporter>>,
}

impl HookMonitor {
    /// Run the monitor loop until shutdown.
    pub async fn run(self, mut shutdown: watch::Receiver<bool>) {
        let sleep_time = Duration::from_secs(self.cfg.sleep_time.max(1) as u64);
        let retry_time = Duration::from_secs(self.cfg.retry_time.max(1) as u64);
        let mut fail_count: i64 = 0;
        let mut was_down = false;

        loop {
            let ok = self.check().await;
            let wait = if ok {
                fail_count = 0;
                if was_down {
                    self.failed.store(false, Ordering::Relaxed);
                    was_down = false;
                    eprintln!("[hook] main DNS recovered, switching back to main DNS");
                    if let Some(rep) = &self.pplog {
                        rep.report_event(SEVERITY_INFO, "[hook] main DNS recovered");
                    }
                    spawn_exec(self.cfg.switch_main_exec.clone());
                }
                sleep_time
            } else {
                fail_count += 1;
                if fail_count >= self.cfg.count && !was_down {
                    self.failed.store(true, Ordering::Relaxed);
                    was_down = true;
                    self.cache.flush();
                    eprintln!(
                        "[hook] main DNS marked DOWN ({fail_count} failures), switching to fallback; cache flushed"
                    );
                    if let Some(rep) = &self.pplog {
                        rep.report_event(
                            SEVERITY_WARN,
                            "[hook] main DNS DOWN, switching to fallback",
                        );
                    }
                    // switch_fall_exec runs after retry_time/2 so the fallback is
                    // already active — and only if still down at that point.
                    if !self.cfg.switch_fall_exec.is_empty() {
                        let cmd = self.cfg.switch_fall_exec.clone();
                        let delay = retry_time / 2;
                        let failed = self.failed.clone();
                        tokio::spawn(async move {
                            tokio::time::sleep(delay).await;
                            if failed.load(Ordering::Relaxed) {
                                run_exec_cmd(&cmd).await;
                            }
                        });
                    }
                }
                retry_time
            };

            tokio::select! {
                _ = shutdown.changed() => {
                    if *shutdown.borrow() { break; }
                }
                _ = tokio::time::sleep(wait) => {}
            }
        }
    }

    /// Probe the primary DNS. With neither `exit_code` nor `keyword` configured,
    /// success is exit 0; otherwise both configured checks must pass.
    async fn check(&self) -> bool {
        let Some((exit_code, output)) = run_shell(&self.cfg.exec, HOOK_CHECK_TIMEOUT).await else {
            return false; // spawn error or timeout
        };
        if !self.cfg.exit_code_set && self.cfg.keyword.is_empty() {
            return exit_code == 0;
        }
        if self.cfg.exit_code_set && exit_code != self.cfg.exit_code {
            return false;
        }
        if !self.cfg.keyword.is_empty() && !contains_sub(&output, self.cfg.keyword.as_bytes()) {
            return false;
        }
        true
    }
}

fn shell() -> String {
    std::env::var("SHELL").unwrap_or_else(|_| "/bin/sh".to_string())
}

/// Run `cmd_str` via the shell. Returns `(exit_code, combined stdout+stderr)`,
/// or None on spawn error / timeout. Output is bounded to `MAX_HOOK_OUTPUT`
/// *during* reading (excess is drained and discarded) so a runaway command
/// cannot grow the buffer without limit.
async fn run_shell(cmd_str: &str, timeout: Duration) -> Option<(i32, Vec<u8>)> {
    let mut cmd = Command::new(shell());
    cmd.arg("-c")
        .arg(cmd_str)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true);
    let mut child = cmd.spawn().ok()?;
    let stdout = child.stdout.take()?;
    let stderr = child.stderr.take()?;

    let run = async {
        // Read both pipes concurrently (bounded), then reap the child.
        let (o, e) = tokio::join!(read_capped(stdout), read_capped(stderr));
        let status = child.wait().await.ok()?;
        let mut combined = o;
        combined.extend_from_slice(&e);
        combined.truncate(MAX_HOOK_OUTPUT);
        Some((status.code().unwrap_or(-1), combined))
    };
    tokio::time::timeout(timeout, run).await.ok()?
}

/// Read an async stream into a buffer, retaining at most `MAX_HOOK_OUTPUT`
/// bytes but continuing to drain the rest so the child never blocks on a full
/// pipe.
async fn read_capped<R: AsyncReadExt + Unpin>(mut r: R) -> Vec<u8> {
    let mut buf = Vec::new();
    let mut chunk = [0u8; 8192];
    loop {
        match r.read(&mut chunk).await {
            Ok(0) | Err(_) => break,
            Ok(n) => {
                if buf.len() < MAX_HOOK_OUTPUT {
                    let take = (MAX_HOOK_OUTPUT - buf.len()).min(n);
                    buf.extend_from_slice(&chunk[..take]);
                }
                // Bytes beyond the cap are read but discarded.
            }
        }
    }
    buf
}

/// Run a side-effect command (switch_*_exec), logging failures.
async fn run_exec_cmd(cmd_str: &str) {
    if cmd_str.is_empty() {
        return;
    }
    match run_shell(cmd_str, HOOK_CMD_TIMEOUT).await {
        Some((0, _)) => {}
        Some((code, out)) => eprintln!(
            "[hook] switch exec failed (exit {code}): {}",
            String::from_utf8_lossy(&out)
        ),
        None => eprintln!("[hook] switch exec failed to run or timed out: {cmd_str}"),
    }
}

fn spawn_exec(cmd_str: String) {
    if cmd_str.is_empty() {
        return;
    }
    tokio::spawn(async move { run_exec_cmd(&cmd_str).await });
}

fn contains_sub(haystack: &[u8], needle: &[u8]) -> bool {
    if needle.is_empty() {
        return true;
    }
    haystack.windows(needle.len()).any(|w| w == needle)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn monitor(cfg: HookConfig) -> HookMonitor {
        HookMonitor {
            cfg,
            failed: Arc::new(AtomicBool::new(false)),
            cache: Arc::new(Cache::new(16)),
            pplog: None,
        }
    }

    fn cfg(exec: &str) -> HookConfig {
        HookConfig {
            exec: exec.to_string(),
            ..HookConfig::default()
        }
    }

    #[tokio::test]
    async fn exit_zero_is_success_by_default() {
        assert!(monitor(cfg("true")).check().await);
        assert!(!monitor(cfg("false")).check().await);
    }

    #[tokio::test]
    async fn keyword_must_be_present() {
        let mut ok = cfg("printf hello-world");
        ok.keyword = "hello".to_string();
        assert!(monitor(ok).check().await);

        let mut no = cfg("printf hello-world");
        no.keyword = "absent".to_string();
        assert!(!monitor(no).check().await);
    }

    #[tokio::test]
    async fn exit_code_checked_only_when_set() {
        let mut ok = cfg("exit 3");
        ok.exit_code = 3;
        ok.exit_code_set = true;
        assert!(monitor(ok).check().await);

        let mut no = cfg("exit 3");
        no.exit_code = 0;
        no.exit_code_set = true;
        assert!(!monitor(no).check().await);
    }

    #[tokio::test]
    async fn spawn_failure_is_treated_as_down() {
        // A command that cannot run (empty → sh error / nonzero) is not success.
        assert!(
            !monitor(cfg("this-command-does-not-exist-xyz"))
                .check()
                .await
        );
    }
}
